lib/fog/openstack/models/image_v2/image.rb

```diff
@@ -66,27 +71,44 @@ def create
             self
           end
 
-          # This overrides the behaviour of Fog::OpenStack::Model::save() which tries to be clever and
-          #  assumes save=update if an ID is present - but Image V2 allows ID to specified on creation
-          def identity
-            nil
-          end
-
-          # Hash of attributes to update is passed in. Setting value to nil will delete that attribute.
-          #  Here we convert that hash into a form suitable for Glance's usage of JSON Patch (RFC6902)
-          def update(attr = {})
+          # Here we convert 'attributes' into a form suitable for Glance's usage of JSON Patch (RFC6902).
+          #  We fetch the existing attributes from the server to compute the delta (difference)
+          #  Setting value to nil will delete that attribute from the server.
+          def update(attr = nil)
             requires :id
+            client_attributes = attr || @attributes
+            server_attributes = service.images.get(id).attributes
+
+            attributes_to_remove = client_attributes.select { |key, value| value.nil? && !server_attributes[key].nil? }
+            attributes_to_add = client_attributes.reject { |key, _| server_attributes.key?(key) }
+            attributes_to_replace = client_attributes.reject { |key, value| value.nil? || server_attributes[key] == value }
+
             json_patch = []
-            attr.each do |key, value|
-              op = (@attributes.keys.include? key) ? 'replace' : 'add'
-              op = 'remove' if value.nil?
-              json_patch << {:op => op, :path => "/#{key}", :value => value}
+            attributes_to_remove.each do |key, value|
+              json_patch << {:op => 'remove', :path => "/#{key}", :value => value}
             end
+            attributes_to_add.each do |key, value|
+              json_patch << {:op => 'add', :path => "/#{key}", :value => value}
+            end
+            attributes_to_replace.each do |key, value|
+              json_patch << {:op => 'replace', :path => "/#{key}", :value => value}
+            end
+
             merge_attributes(
               service.update_image(id, json_patch).body)
             self
           end
 
+          # This overrides the behaviour of Fog::OpenStack::Model::save() which tries to be clever and
+          #  assumes save=update if an ID is present - but Image V2 allows ID to specified on creation
+          def save
+            create
+          rescue StandardError => e
+            raise e unless e.response && e.response.body && e.response.body.match(/Attribute .* is read-only./)
```

Comment:
I would have expected the API to return a 409 Conflict if the resource already exists, in fact it will under some circumstances. But in our test cases for Image V2, calling `create` if the ID already exists instead returned 403 "Attribute 'self' is read-only". We'll open an issue on the OpenStack API to look into this further.